### ISSUE TITLE
Freeze layers with the multi_gpu_model in Keras

### ISSUE THREAD

**Issue by chaohuang** [keras-team/keras#9198]
I'm trying to fine-tune a modified InceptionV3 model in Keras.

I follow the example "Fine-tune InceptionV3 on a new set of classes" on [this page][1].

So I first trained the top dense layers that were added to the InceptionV3 base model with the following code:

    model = Model(inputs=base_model.input, outputs=predictions)

    for layer in base_model.layers:
        layer.trainable = False

    parallel_model = multi_gpu_model(model, gpus=2)
    
    parallel_model.compile(optimizer='rmsprop', loss='categorical_crossentropy')
    
    history = parallel_model.fit_generator(generate_batches(path), steps_per_epoch = num_images/batch_size, epochs = num_epochs)

After that, I try to fine-tune the top 2 inception blocks from InceptionV3. And according to the example, what I should do is:

    for layer in model.layers[:249]:
       layer.trainable = False
    for layer in model.layers[249:]:
       layer.trainable = True
    
    model.compile(optimizer=SGD(lr=0.0001, momentum=0.9), loss='categorical_crossentropy')

    model.fit_generator(...)

But I'm using the `multi_gpu_model`, so I don't know how to freeze the first 249 layers. 

I mean, if I freeze the layers in the no-gpu model (like the example), and use `parallel_model = multi_gpu_model(model, gpus=2)` to freeze the layers in the `parallel_model`, then the weights in the top dense layers that were just trained and contained in the `parallel_model` will be overwritten, right?

On the other hand, I tried to directly use `for layer in parallel_model.layers[:249]: layer.trainable = False`, but when I checked the layers in the `parallel_model`, it showed:

    for i, layer in enumerate(parallel_model.layers):
       print(i, layer.name)

    (0, 'input_1')
    (1, 'lambda_1')
    (2, 'lambda_2')
    (3, 'model_1')
    (4, 'dense_3')

So what are the 'lambda_1', 'lambda_2' and 'model_1' layers and why it only shows 5 layers in the `parallel_model`?

More importantly, how to freeze the layers in the `parallel_model`?

  [1]: https://keras.io/applications/

**Comment by rex-yue-wu** [comment]
I think you actually did freeze your `parallel_model` (PM) through your single-gpu model (SM). 

So if you do anything w.r.t. this SM, it will be reflected in the corresponding PM. This means if you freeze `K` parameters in SM, you will also see `K` non-trainable parameters in PM.  

More specific, if you do something below
```
single_model = Model(inputs=base_model.input, outputs=predictions)

# freeze/unfreeze weights in SM
for layer in single_model.layers:
    layer.trainable = False

parallel_model = multi_gpu_model(single_model, gpus=2)
parallel_model.compile(optimizer='rmsprop', loss='categorical_crossentropy')

print single_model.summary()
print parallel_model.summary()
```
you will see both print out messages show the same number of non-trainable parameters.

Once again, you should only freeze `a layer` when it is a `keras layer` (just as SM in this case, it is actually `model`)

**Comment by fchollet** [comment]
This is correct. You can set `trainable` on layers from the original `model`, and that will be taken into account in `parallel_model`. 

However, you should always remember to `compile` `parallel_model` after changing any trainable value. Compiling is required for any changes to be taken into account.

**Comment by chaohuang** [comment]
Thanks, but let me make sure I understand correctly.

Let's say, I first freeze some layers in the original `model`, then call `parallel_model = multi_gpu_model(model, gpus=2)`, `parallel_model.compile(...)` and `parallel_model.fit(...)`. I understand this because the statement `parallel_model = multi_gpu_model(model, gpus=2)` will copy the `trainable` setting transferred from the original `model` to the `parallel_model`.

But after that, I want to freeze other layers in the `parallel_model`. So this time, I only need to set trainable on layers from the original `model` and compile the `parallel_model`, and I do NOT need to call `parallel_model = multi_gpu_model(model, gpus=2)` to freeze the layers in the `parallel_model`, is that correct?

If so, how is the `trainable` setting transferred from the original `model` to the `parallel_model` in this case?

**Comment by fchollet** [comment]
Yes, that's correct. As for "transfer", you must understand that they share
the same layers, so there is nothing to transfer.
